BuildSourceImage.sh

```diff
@@ -1,80 +1,968 @@
+    fi
+
+    ## TODO: check for authfile, creds, and whether it's an insecure registry
+    ## destination name must have the image tag included (umoci expects it)
+    skopeo \
+        copy \
+        "${from}" \
+        "oci:${dst}:${tag}" >&2
+    echo -n "${dst}:${tag}"
+}
+
+#
+# upack_img <oci layout path> <unpack path>
+#
+unpack_img() {
+    local image_dir="${1}"
+    local unpack_dir="${2}"
+
+    if [ -d "${unpack_dir}" ] ; then
+        _rm_rf "${unpack_dir}"
+    fi
+
+    # TODO perhaps if uid == 0 and podman is present then we can try it?
+    if [ -z "$(command -v umoci)" ] ; then
+        # can be done as non-root (even in a non-root container)
+        unpack_img_umoci "${image_dir}" "${unpack_dir}"
+    else
+        # can be done as non-root (even in a non-root container)
+        unpack_img_bash "${image_dir}" "${unpack_dir}"
+    fi
+}
+
+#
+# unpack an image layout using only jq and bash
+#
+unpack_img_bash() {
+    local image_dir="${1}"
+    local unpack_dir="${2}"
+
+    # for compat with umoci (which wants the image tag as well)
+    if echo "${image_dir}" | grep -q ":" ; then
+        image_dir="${image_dir%:*}"
+    fi
+
+    local mnfst_dgst="$(cat "${image_dir}"/index.json | jq '.manifests[0].digest' | tr -d \" )"
+
+    # Since we're landing the reference as an OCI layout, this mediaType is fairly predictable
+    # TODO don't always assume +gzip
+    layer_dgsts="$(cat ${image_dir}/blobs/${mnfst_dgst/:/\/} | \
+        jq '.layers[] | select(.mediaType == "application/vnd.oci.image.layer.v1.tar+gzip") | .digest' | tr -d \")"
+
+    _mkdir_p "${unpack_dir}/rootfs"
+    for dgst in ${layer_dgsts} ; do
+        path="${image_dir}/blobs/${dgst/:/\/}"
+        tmp_file=$(_mktemp)
+        zcat "${path}" | _tar -t > $tmp_file # TODO cleanup these files
+
+        # look for '.wh.' entries. They must be removed from the rootfs
+        # _before_ extracting the archive, then the .wh. entries themselves
+        # need to not remain afterwards
+        grep '\.wh\.' "${tmp_file}" | while read line ; do
+            # if `some/path/.wh.foo` then `rm -rf `${unpack_dir}/some/path/foo`
+            # if `some/path/.wh..wh..opq` then `rm -rf `${unpack_dir}/some/path/*`
+            if [ "$(basename ${line})" == ".wh..wh..opq" ] ; then
+                _rm_rf "${unpack_dir}/rootfs/$(dirname ${line})/*"
+            elif basename "${line}" | grep -qe '^\.wh\.' ; then
+                name=$(basename "${line}" | sed -e 's/^\.wh\.//')
+                _rm_rf "${unpack_dir}/rootfs/$(dirname ${line})/${name}"
+            fi
+        done
+
+        _info "[unpacking] layer ${dgst}"
+        # unpack layer to rootfs (without whiteouts)
+        zcat "${path}" | _tar --restrict --no-xattr --no-acls --no-selinux --exclude='*.wh.*' -x -C "${unpack_dir}/rootfs"
+
+        # some of the directories get unpacked as 0555, so removing them gives an EPERM
+        find "${unpack_dir}" -type d -exec chmod 0755 "{}" \;
+    done
+}
```

Comment:
Do we need to treat exit codes in some way here as we do in `unpack_img_umoci()`?

Reply:
yea. This needs to be more thorough throughout.
```bash
ret=$?
if [ ${ret} -ne 0 ] ; then
  return ${ret}
fi
```